documentation.md

```diff
@@ -0,0 +1,521 @@
+
+- The delay calculator, which has as major responsibility to look at the request traffic and to deduce, according to the simulated memory controller parameters.
+- The response banks, which have as major responsibilities to:
+  a. Store the responses from the real memory controller until they can be released.
+  b. Enforce the ordering per AXI identifier, letting the delay calculator be agnostic of the AXI identifier.
+
+The delay calculator identifies address requests by internal identifiers (_iids_), which are not correlated with AXI identifiers. Precisely, write responses and read data use separate iid spaces: each response to an address request is therefore identified by the pair (read/write, iid), although the first entry of the pair is always implicit.
+
+<figure class="image">
+  <img src="https://i.imgur.com/ri0Ej2o.png" alt="Simulated memory controller internal overview">
+  <figcaption>Fig: Simulated memory controller overview</figcaption>
+</figure>
+
+#### Top-level flow
+
+The top-level flow happens as follows:
+
+0. The response banks constantly advertise the identifier that will be allocated to the next address request.
+1. When the address request comes in:
+   - The delay calculator and the response banks are notified when an address request happens.
+   - The delay calculator stores metadata for its delay computation.
+   - The response banks reserve space for the corresponding responses that will subsequently come back from the real memory controller.
+2. When the corresponding responses comes back from the real memory controller, they are stored by the message banks.
+3. When the delay enables the release of some response and all the previous responses of the same AXI are released, the response banks transmit the response to the requester.
+4. When all the responses corresponding to a given address request have been transmitted to the requester, the response banks free the allocated memory for reuse.
+5.
+
+<figure class="image">
+  <img src="https://i.imgur.com/pbbpy0Y.png" alt="Simulated memory controller internal overview">
+  <figcaption>Fig: Simulated memory controller top-level flow</figcaption>
+</figure>
+
+## Response banks
+
+### High-level design
+
+The _simmem_resp_banks_ module is made of two distinct _simmem_resp_bank_ modules:
+
+- The write response bank, responsible for storing write responses.
+- The read data bank, responsible for storing read data. As read data can come back as burst data, this bank has to additionally support bursts, which is the essential difference with the write response bank.
+
+The two banks act independently.
+
+Each response bank has a FIFO interface with reservations, and stores messages in RAMs. FIFOs are implemented as linked lists to make best use of the RAM space.
+
+<figure class="image">
+  <img src="https://i.imgur.com/BpU4aZz.png" alt="Simulated memory controller internal overview">
+  <figcaption>Fig: Response bank top-level interface</figcaption>
+</figure>
```

Comment:
The figure is slightly misleading as it suggest that there is a separate structure/FIFO for every AXI ID. But in fact, you have just a couple of registers per ID to store the pointers into the shared RAMs. I wonder if it is somehow possible to convey this information in that figure.

Reply:
Alright! I put that in the caption.